ocd_backend/pipeline.py

```diff
@@ -52,46 +53,113 @@ def setup_pipeline(source_definition):
         'run_identifier': 'pipeline_{}'.format(uuid4().hex),
         'current_index_name': current_index_name,
         'new_index_name': new_index_name,
-        'index_alias': index_alias,
-        'source_definition': source_definition
+        'index_alias': index_alias
     }
 
     celery_app.backend.set(params['run_identifier'], 'running')
     run_identifier_chains = '{}_chains'.format(params['run_identifier'])
 
-    first_step = source_definition['chain'].pop(0)
-    try:
-        for item in load_object(first_step)(**params).run():
-            step_chain = chain()
-
-            for step in source_definition['chain']:
-                # Generate an identifier for each chain, and record that in
-                # {}_chains, so that we can know for sure when all tasks
-                # from an extractor have finished
+    # we can have multiple pipelines. but for compatibility and readability
+    # use the source definition if no specific pipelines have been defined
+    pipelines = source_definition.get('pipelines', None) or [source_definition]
+
+    pipeline_definitions = {}
+    pipeline_extractors = {}
+    pipeline_transformers = {}
+    pipeline_enrichers = {}
+    pipeline_loaders = {}
+
+    for pipeline in pipelines:
+        if 'id' not in pipeline:
+            raise ConfigurationError("Each pipeline must have an id field.")
+
+        # adjusted source definitionsv per pipeline. This way you can for
+        # example change the index on a pipeline basis
+        pipeline_definitions[pipeline['id']] = deepcopy(source_definition)
+        pipeline_definitions[pipeline['id']].update(pipeline)
+
+        # initialize the ETL classes, per pipeline
+        pipeline_extractors[pipeline['id']] = [
+            load_object(cls) for cls in
```

Comment:
Classes are not instantiated here (which does happen for transformers, enrichers, loaders) since we are dealing with both generator and task extractors. The first extractor (the generator) should take a source_definition (line 107) and possible extractor tasks (118) take nothing. I saw no other way around than to instantiate on those lines.